coaster/views.py

```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
                 return result
         return decorated_function
     return inner
+
+def cors(is_origin_allowed=lambda origin: False,
+    allowed_methods='HEAD, OPTIONS, GET, POST, PUT, PATCH, DELETE',
+    allowed_headers='Accept, Accept-Language, Content-Language, Content-Type, X-Requested-With',
+    max_age=None):
+    """
+    Adds CORS headers to the decorated view function.
+
+    :param is_origin_allowed: A handler function that receives the origin as a parameter and
+    is expected to return a boolean value to assert if the given origin has access to the
+    requested resource
+    :param allowed_methods: A string of comma-separated HTTP methods that are allowed for this origin
+    :param allowed_headers: A string of comma-separated HTTP headers that are allowed for this origin
+    :param max_age: Maximum number of seconds the result for the pre-flight request can be cached
+    """
+    @wraps(is_origin_allowed)
+    def inner(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            origin = request.headers.get('Origin')
+            if request.method not in allowed_methods or not is_origin_allowed(origin):
+                abort(401)
+
+            if request.method == 'OPTIONS':
+                # pre-flight request
+                resp = Response()
+            else:
+                resp = f(*args, **kwargs)
```

Comment:
The view may not return a response object. Flask allows views to return any of the following:

1. A response object
2. A string (response body)
3. A 2-tuple consisting of a string and a status code
4. A 3-tuple consisting of a string, status code and HTTP headers

See `coaster.views.render_with` for how we handle these variations.